### DISCUSSION TITLE
How to fetch a user's permission level using GitHub App and GraphQL?

### DISCUSSION THREAD

**Discussion by wata727** [community/community#24513] [Apps, API and Webhooks]
<p>I’m trying to fetch <a href="https://developer.github.com/v3/repos/collaborators/#review-a-users-permission-level" rel="nofollow noopener">a user’s permission level</a> using GitHub App and GraphQL. However, there seems to be only viewerPermission in the attributes of the repository. This value is always null in GitHub App.</p>
<p>Can I write queries that add permissions for the specific user in a repository list?</p>

**Comment by ghost** [comment]
<p>Hey there <a class="mention" href="/wata727">@wata727</a>, help’s arrived!</p>
<p>If you want, GitHub provides an API endpoint for adding different users as collaborators on a repository: <a href="https://developer.github.com/v3/repos/collaborators/#add-user-as-a-collaborator" rel="nofollow noopener">https://developer.github.com/v3/repos/collaborators/#add-user-as-a-collaborator</a>. Is this what you’re looking for? 🙂</p>

**Comment by wata727** [comment]
<p>Hello <a class="mention" href="/timrossback">@timrossback</a>, Thanks for your reply!</p>
<p>Unfortunately, that is not the API I want…</p>
<p>I would like to list repositories including a user’s permission like the following:</p>
<pre><code>{
  "data": {
    "organization": {
      "repositories": {
        "edges": [
          {
            "node": {
              "name": "transfer",
              "viewerPermission": "ADMIN"
            }
          },
          {
            "node": {
              "name": "sandbox_fuelphp",
              "viewerPermission": "ADMIN"
            }
          },
          {
            "node": {
              "name": "php_sandbox",
              "viewerPermission": "ADMIN"
            }
          }
        ]
      }
    }
  }
}
</code></pre>
<p>The above response can be obtained by the following query:</p>
<pre><code>query {
  organization(login: "wata727-sandbox") {
    repositories(first: 3) {
      edges {
        node {
          name
          viewerPermission
        }
      }
    }
  }
}
</code></pre>
<p>However, viewPermission is always null in GitHub App like the following:</p>
<pre><code>{
  "data": {
    "organization": {
      "repositories": {
        "edges": [
          {
            "node": {
              "name": "transfer",
              "viewerPermission": null
            }
          },
          {
            "node": {
              "name": "sandbox_fuelphp",
              "viewerPermission": null
            }
          },
          {
            "node": {
              "name": "php_sandbox",
              "viewerPermission": null
            }
          }
        ]
      }
    }
  }
}
</code></pre>
<p>I want to avoid this and know how to fetch permission for a specific user. GitHub API v3 provides <a href="https://developer.github.com/v3/repos/collaborators/#review-a-users-permission-level" rel="nofollow noopener">permission level API</a> but this API requires to call it for each repository. This is very expensive…</p>

**Comment by ghost** [comment]
<p>Hey <a class="mention" href="/wata727">@wata727</a>,</p>
<p>I get what you mean! I’m probably not the best person to talk to about this since I don’t actively develop anything that integrates with GitHub. You might find that Stack Overflow might be able to help, though. Otherwise, you can always try asking the wonderful folks over at <a href="https://github.com/contact" rel="nofollow noopener">GitHub support</a> :slightly_smiling_face:</p>

**Comment by mathstuf** [comment]
<p>The field you want is on the *edge* of collaborators. See <a>this object’s <code>permission</code> field</a>.</p>

**Comment by dione2** [comment]
<p>It will remain null for anyone other than the owner</p>

**Comment by Nosferican** [comment]
<p>Would be great to have a way to query the permission type given some user and the repository.</p>

**Comment by ilia-cy** [comment, marked as answer]
<p>This API does exist:</p>
<p><a href="https://api.github.com/repos/OWNER/REPO/collaborators/USERNAME/permission" class="onebox" target="_blank" rel="nofollow noopener">https://api.github.com/repos/OWNER/REPO/collaborators/USERNAME/permission</a></p>
<p>The response is the following:</p>
<pre><code class="lang-auto">{
    "permission": "admin",
    "user": {
        "login": "USERNAME",
        "id": 1,
        "node_id": "NODEID",
        "avatar_url": "https://avatars1.githubusercontent.com/u/",
        "gravatar_id": "",
        "url": "https://api.github.com/users/USERNAME",
        "html_url": "https://github.com/USERNAME",
        "followers_url": "https://api.github.com/users/USERNAME/followers",
        "following_url": "https://api.github.com/users/USERNAME/following{/other_user}",
        "gists_url": "https://api.github.com/users/USERNAME/gists{/gist_id}",
        "starred_url": "https://api.github.com/users/USERNAME/starred{/owner}{/repo}",
        "subscriptions_url": "https://api.github.com/users/USERNAME/subscriptions",
        "organizations_url": "https://api.github.com/users/USERNAME/orgs",
        "repos_url": "https://api.github.com/users/USERNAME/repos",
        "events_url": "https://api.github.com/users/USERNAME/events{/privacy}",
        "received_events_url": "https://api.github.com/users/USERNAME/received_events",
        "type": "User",
        "site_admin": false
    }
}
</code></pre>
